crates/node-core/src/init.rs

```diff
@@ -233,6 +272,115 @@ pub fn insert_genesis_header<DB: Database>(
     Ok(())
 }
 
+/// Initialize chain with state at specific block, from reader of state dump.
+pub fn init_from_state_dump<DB: Database>(
+    mut reader: impl BufRead,
+    factory: ProviderFactory<DB>,
+) -> eyre::Result<B256> {
+    let block = factory.last_block_number()?;
+    let hash = factory.block_hash(block)?.unwrap();
+
+    debug!(target: "reth::cli",
+        block,
+        chain=%factory.chain_spec().chain,
+        "Initializing state at block"
+    );
+
+    let mut total_inserted_accounts = 0;
+    let mut accounts = Vec::with_capacity(DEFAULT_LEN_ACCOUNTS_CHUNK);
+    let mut line = String::new();
+
+    // first line can be state root, then it can be used for verifying against computed state root
+    reader.read_line(&mut line)?;
+    let _state_root = match serde_json::from_str::<StateRoot>(&line) {
+        Ok(root) => {
+            trace!(target: "reth::cli",
+                root=%root.root,
+                "Read state root from file"
+            );
+            Some(root)
+        }
+        Err(_) => {
+            let GenesisAccountWithAddress { genesis_account, address } =
+                serde_json::from_str(&line)?;
+            accounts.push((address, genesis_account));
+
+            None
+        }
+    };
+    line.clear();
+
+    // remaining lines are accounts
+    while let Ok(n) = reader.read_line(&mut line) {
+        if accounts.len() == DEFAULT_LEN_ACCOUNTS_CHUNK || n == 0 {
+            total_inserted_accounts += accounts.len();
+            debug!(target: "reth::cli",
+                block,
+                parsed_new_accounts=accounts.len(),
+                total_inserted_accounts,
+                "Writing accounts to db"
+            );
+            // use transaction to insert genesis header
+            let provider_rw = factory.provider_rw()?;
+            insert_genesis_hashes(
+                &provider_rw,
+                accounts.iter().map(|(address, account)| (address, account)),
+            )?;
+            insert_history(
+                &provider_rw,
+                accounts.iter().map(|(address, account)| (address, account)),
+                block,
+            )?;
+
+            // block is already written to static files
+            let tx = provider_rw.into_tx();
+
+            insert_state::<DB>(
+                &tx,
+                accounts.len(),
+                accounts.iter().map(|(address, account)| (address, account)),
+                block,
+                false,
+            )?;
+
+            // insert sync stage
+            for stage in StageId::ALL.iter() {
+                tx.put::<tables::StageCheckpoints>(stage.to_string(), StageCheckpoint::new(block))?;
```

Comment:
yeah I think so. necessary to start syncing from that given block, no? it's done in `init_genesis` but for block 0

Reply:
ah I see, the state is for one block, so this should happen once. but will remove it since state import assumes the chain segment until the block that is intended to import state for, is imported. otherwise users should default to init cmd, is the idea.